### PR TITLE
[ADD] Added new module two_factor_otp_auth

### PR DESCRIPTION
**Pull request by Werded**

### CODE REVIEW

File: two_factor_otp_auth/README.rst [hunk 1]
```diff
@@ -0,0 +1,19 @@
+================================
+Ventor Two Factor Authentication
+================================
+
+
+
+
+Credits
```
Comment (okuryan) [inline two_factor_otp_auth/README.rst]: @Werded , please, add readme file similar to the https://github.com/OCA/mis-builder/blob/10.0/mis_builder/README.rst 
Note that some sections are not needed, of course.

File: two_factor_otp_auth/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2019 VentorTech OU
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0).
+
+{
+    "name": "Ventor Two Factor Authentication",
+    "summary": "Two factor authentication for users",
+    "author": "VentorTech OU",
+    "website": "http://www.ventor.tech/",
```
Comment (okuryan) [inline two_factor_otp_auth/__manifest__.py]: Site should be "https://ventor.tech"

File: two_factor_otp_auth/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2019 VentorTech OU
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0).
+
+{
+    "name": "Ventor Two Factor Authentication",
+    "summary": "Two factor authentication for users",
+    "author": "VentorTech OU",
```
Comment (okuryan) [inline two_factor_otp_auth/__manifest__.py]: Author should be as in other merp modules "VentorTech"

File: two_factor_otp_auth/controllers/main.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    def web_login(self, redirect=None, **kw):
+        """
+        Overload core method to start Second Factor validation step
+        """
+        try:
+            response = super(Login2fa, self).web_login(redirect, **kw)
+        except MissingOtpError:
+            # user will coming here while login process is not full successful
+            # come here after success validation of other credentials
+            # to start Second Factor (OTP) validation step
```
Comment (okuryan) [inline two_factor_otp_auth/controllers/main.py]: Fix some typos in comment:

 user will get into this block if login process is not fully successful
 For example, when first login was successful, but 2FA token is missing
 So we can start second authentication step (OTP)

File: two_factor_otp_auth/controllers/main.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                        "secret_code_2fa": params.get("secret_code_2fa"),
+                    }
+                    user.sudo().write(values)
+
+        return response
+
+    @staticmethod
+    def _get_response(message=None):
+        """
+        Method for getting response object which depending on user and values
```
Comment (okuryan) [inline two_factor_otp_auth/controllers/main.py]: Correct grammar mistakes: Method to get response object that depends on user and request params values

File: two_factor_otp_auth/data/ir_actions_server_data.xml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+        <record id="ir_actions_server_user_enable_2fa" model="ir.actions.server">
+            <field name="name">Enable Two Factor Auth</field>
+            <field name="model_id" ref="model_res_users"/>
+            <field name="binding_model_id" ref="model_res_users"/>
+            <field name="binding_type">action</field>
+            <field name="state">code</field>
+            <field name="activity_user_type">specific</field>
+            <field name="code">action = records.action_enable_2f_auth()</field>
+        </record>
```
Comment (okuryan) [inline two_factor_otp_auth/data/ir_actions_server_data.xml:12]: @Werded , please, make sure that both this actions are visible on for user with Technical Settings.

File: two_factor_otp_auth/exceptions/access_errors.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2019 VentorTech OU
```
Comment (okuryan) [inline two_factor_otp_auth/exceptions/access_errors.py]: Change everywhere to 2020 =)

File: two_factor_otp_auth/models/res_users.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        key = b32encode(urandom(10))
+        code = pyotp.totp.TOTP(key).provisioning_uri(self.login)
+        img = qrcode.make(code)
+        _, file_path = mkstemp()  # creating temporary file
+        img.save(file_path)
+
+        with open(file_path, 'rb') as image_file:
+            qr_image_code = b64encode(image_file.read())
+
+        remove(file_path)  # removing temporary file
```
Comment (okuryan) [inline two_factor_otp_auth/models/res_users.py]: https://stackoverflow.com/questions/10840533/most-pythonic-way-to-delete-a-file-which-may-not-exist/27045091#27045091 - removing file there are multiple ways. Check 2 methods - unlink and remove.

File: two_factor_otp_auth/models/res_users.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        Generate QR-Code based on random set of letters
+
+        Returns:
+         * tuple - generated secret_code and binary qr-code
+        """
+        self.ensure_one()
+        key = b32encode(urandom(10))
+        code = pyotp.totp.TOTP(key).provisioning_uri(self.login)
+        img = qrcode.make(code)
+        _, file_path = mkstemp()  # creating temporary file
```
Comment (okuryan) [inline two_factor_otp_auth/models/res_users.py:130]: I'm a little worried that you create file, open it and delete it in different blocks. When you work with OS, there can happen different errors. Maybe make sense to do it in try / except / final block? Please, investigate this point.

File: two_factor_otp_auth/models/res_users.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+           action
+
+        TODO:
+         * Rewrite text of warning - add list of groups with access
+         * Or even better create separate group.
+        """
+        if not self.env.user._is_admin():
+            raise AccessError(_(
+                "Only Administrators can do this operation!"
+            ))
```
Comment (okuryan) [inline two_factor_otp_auth/models/res_users.py:158]: Raise exception if user do not belong to the group "Technical Settings".

Remove TODOs when ready

**Review by okuryan — COMMENTED** [review]
Also make sure that in merp_base added checkbox in settings that can auto-install this module. As is done with all other modules

File: two_factor_otp_auth/models/res_users.py (hunk 10 above)
Comment (Werded) [reply two_factor_otp_auth/models/res_users.py:158]: done

File: two_factor_otp_auth/models/res_users.py (hunk 9 above)
Comment (Werded) [reply two_factor_otp_auth/models/res_users.py:130]: done

File: two_factor_otp_auth/models/res_users.py (hunk 8 above)
Comment (Werded) [reply two_factor_otp_auth/models/res_users.py]: done

File: two_factor_otp_auth/exceptions/access_errors.py (hunk 7 above)
Comment (Werded) [reply two_factor_otp_auth/exceptions/access_errors.py]: done

File: two_factor_otp_auth/data/ir_actions_server_data.xml (hunk 6 above)
Comment (Werded) [reply two_factor_otp_auth/data/ir_actions_server_data.xml:12]: done

File: two_factor_otp_auth/controllers/main.py (hunk 5 above)
Comment (Werded) [reply two_factor_otp_auth/controllers/main.py]: done

File: two_factor_otp_auth/controllers/main.py (hunk 4 above)
Comment (Werded) [reply two_factor_otp_auth/controllers/main.py]: done

File: two_factor_otp_auth/__manifest__.py (hunk 3 above)
Comment (Werded) [reply two_factor_otp_auth/__manifest__.py]: done

File: two_factor_otp_auth/__manifest__.py (hunk 2 above)
Comment (Werded) [reply two_factor_otp_auth/__manifest__.py]: done

File: two_factor_otp_auth/README.rst (hunk 1 above)
Comment (Werded) [reply two_factor_otp_auth/README.rst]: done

File: two_factor_otp_auth/README.rst [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+Installation
+~~~~~~~~~~~~
+
+* ``pip install -r <your_addons_path>/merp/two_factor_otp_auth/requirements.txt``
+* then restart Odoo, update the addons list in your database, and install
+
+Authors
+~~~~~~~
+
+* VentorTech OU
```
Comment (okuryan) [inline two_factor_otp_auth/README.rst]: Add "VentorTech OU < hello@ventor.tech > "

File: merp_base/models/res_config.py [hunk 12]
```diff
@@ -48,6 +48,9 @@ class StockConfigSettings(models.TransientModel):
         compute='_compute_merp_version',
         store=False,
     )
+    module_two_factor_otp_auth = fields.Boolean(
+        string='Use Two Factor Authentication'
```
Comment (Tatider) [inline merp_base/models/res_config.py]: Comma at the end of line missed.
https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas

File: two_factor_otp_auth/controllers/main.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+import logging
+
+from odoo import http, _
+from odoo.addons.web.controllers.main import Home
+from odoo.http import request
+
+_logger = logging.getLogger(__name__)
+
+try:
+    from ..exceptions import MissingOtpError, InvalidOtpError
```
Comment (Tatider) [inline two_factor_otp_auth/controllers/main.py]: Is this actually needed if they at the same module?

File: two_factor_otp_auth/controllers/main.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        argument:
+         *message(str) - error message
+
+        Returns:
+         *response object
+        """
+        values = request.params.copy()
+        if message:
+            values.update({
+                'error': message
```
Comment (Tatider) [inline two_factor_otp_auth/controllers/main.py]: 1. trailing comma is missed
2. different types of quotes

File: two_factor_otp_auth/data/ir_actions_server_data.xml [hunk 15]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="1">
+
+        <record id="ir_actions_server_user_enable_2fa" model="ir.actions.server">
+            <field name="name">Enable Two Factor Auth</field>
+            <field name="model_id" ref="model_res_users"/>
+            <field name="binding_model_id" ref="model_res_users"/>
+            <field name="binding_type">action</field>
+            <field name="groups_id" eval="(4, ref('two_factor_otp_auth.mass_change_2fa_for_users'))"/>
```
Comment (Tatider) [inline two_factor_otp_auth/data/ir_actions_server_data.xml]: 1. `groups_id` field for model `ir.actions.server` doesn't exist in [Odoo core](https://github.com/odoo/odoo/blob/12.0/odoo/addons/base/models/ir_actions.py#L321). 
2. `mass_change_2fa_for_users` doesn't exist - file `security/two_factor_otp_auth.xml` doesn't included in `__manifest__.py`
3. '[]' missed in `eval` => `[(4, ref('two_factor_otp_auth.mass_change_2fa_for_users'))]`

File: two_factor_otp_auth/data/ir_actions_server_data.xml [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+            <field name="code">action = records.action_enable_2f_auth()</field>
+
+        </record>
+
+        <record id="ir_actions_server_user_disable_2fa" model="ir.actions.server">
+            <field name="name">Disable Two Factor Auth</field>
+            <field name="model_id" ref="model_res_users"/>
+            <field name="binding_model_id" ref="model_res_users"/>
+            <field name="binding_type">action</field>
+            <field name="groups_id" eval="(4, ref('two_factor_otp_auth.mass_change_2fa_for_users'))"/>
```
Comment (Tatider) [inline two_factor_otp_auth/data/ir_actions_server_data.xml]: The same problem with field and it's future value

File: two_factor_otp_auth/models/res_users.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+            remove(file_path)
+
+        return key, qr_image_code
+
+    @api.model
+    def _can_change_2f_auth_settings(self):
+        """
+        Check that current user can make mass actions with Two Factor
+        Authentication settings.
+        Raises:
```
Comment (Tatider) [inline two_factor_otp_auth/models/res_users.py:150]: Empty line above - over wise will be added to previous line in markdown documentation

File: two_factor_otp_auth/models/res_users.py [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+         * odoo.exceptions.AccessError: only users with "Mass Change 2FA Configuration
+          for Users" rights can do this action
+        """
+        if not self.env.user.has_group('two_factor_otp_auth.mass_change_2fa_for_users'):
```
Comment (Tatider) [inline two_factor_otp_auth/models/res_users.py]: Check outer quotes in this file.

File: two_factor_otp_auth/models/res_users.py [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+        Inverse `enable_2fa` - call `action_discard_2f_auth_credentials` method
+        if value of the field become `false`
+        """
+        for user in self:
+            if not user.enable_2fa:
+                user.action_discard_2f_auth_credentials()
+
+    def action_discard_2f_auth_credentials(self):
+        """
+        Remove values from fields `qr_image_2fa`, `auth_secret_code_2fa`
```
Comment (Tatider) [inline two_factor_otp_auth/models/res_users.py]: `Extra Info` section with explanation why this method does't call `_can_change_2f_auth_settings `

File: two_factor_otp_auth/models/res_users.py [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+            qr_image_code = b64encode(image_file.read())
+
+        # removing temporary file
+        with suppress(OSError):
+            remove(file_path)
+
+        return key, qr_image_code
+
+    @api.model
+    def _can_change_2f_auth_settings(self):
```
Comment (Tatider) [inline two_factor_otp_auth/models/res_users.py]: What about making this one `api.multi` and add check `or self.ids == self.user.ids`

File: two_factor_otp_auth/models/res_users.py [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+        """
+        totp = pyotp.TOTP(secret)
+        str_otp = str(otp)
+        verify = totp.verify(str_otp)
+        if not verify:
+            raise InvalidOtpError()
+        return True
+
+    @api.multi
+    def write(self, vals):
```
Comment (Tatider) [inline two_factor_otp_auth/models/res_users.py]: Why CRUD method (one of most important methods) at the end of file?

File: two_factor_otp_auth/models/res_users.py [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+        """
+        Overload core method to check access rights for changing 2FA
+        """
+        if 'enable_2fa' in vals:
+            self._can_change_2f_auth_settings()
```
Comment (Tatider) [inline two_factor_otp_auth/models/res_users.py]: Should then call of `_can_change_2f_auth_settings` be removed from `action_enable_2f_auth` and `action_disable_2f_auth`

File: two_factor_otp_auth/security/two_factor_otp_auth.xml [hunk 23]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="0">
```
Comment (Tatider) [inline two_factor_otp_auth/security/two_factor_otp_auth.xml]: what the point of `noupdate="0"`?

File: two_factor_otp_auth/views/res_users_view.xml [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+    <record id="res_users_form_inherit_2fa" model="ir.ui.view">
+        <field name="name">res.users.form.inherit.2fa</field>
+        <field name="model">res.users</field>
+        <field name="inherit_id" ref="base.view_users_form"/>
+        <field name="arch" type="xml">
+            <field name="signature" position="before">
+                <field name="enable_2fa"/>
+                <field name="qr_image_2fa"
+                        widget="image"
+                        attrs="{
```
Comment (Tatider) [inline two_factor_otp_auth/views/res_users_view.xml:13]: To improve readability?
```xml
                        attrs="{
                            'invisible': [
                                '|',
                                    ('enable_2fa', '=', False),
                                    ('qr_image_2fa', '=', False),
                            ],
                        }"
```

File: two_factor_otp_auth/views/res_users_view.xml [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+    </record>
+
+    <record id="res_users_tree_inherit_2fa" model="ir.ui.view">
+        <field name="name">res.users.tree.inherit.2fa</field>
+        <field name="model">res.users</field>
+        <field name="inherit_id" ref="base.view_users_tree"/>
+        <field name="arch" type="xml">
+            <field name="login_date" position="after">
+                <field name="enable_2fa"
+                       widget="checkbox"/>
```
Comment (Tatider) [inline two_factor_otp_auth/views/res_users_view.xml]: different types of attribute indents at the same file

File: two_factor_otp_auth/__manifest__.py [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+
+{
+    "name": "Ventor Two Factor Authentication",
+    "summary": "Two factor authentication for users",
+    "author": "VentorTech",
+    "website": "https://ventor.tech",
+    "category": "Uncategorized",
+    "license": "LGPL-3",
+    "version": "12.0.1.0.0",
+    "images": ["static/description/Two_factor_authentification.png"],
```
Comment (Tatider) [inline two_factor_otp_auth/__manifest__.py]: readability?
```python
"images": [
    "static/description/Two_factor_authentification.png",
],
```

File: two_factor_otp_auth/controllers/main.py [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+                    values = {
+                        "qr_image_2fa": params.get("qr_code_2fa"),
+                        "secret_code_2fa": params.get("secret_code_2fa"),
+                    }
+                    user.sudo().write(values)
+
+        return response
+
+    @staticmethod
+    def _get_response(message=None):
```
Comment (Tatider) [inline two_factor_otp_auth/controllers/main.py]: Method name says nothing...

File: two_factor_otp_auth/__manifest__.py [hunk 28]
```diff
@@ -0,0 +1,31 @@
...
+    "installable": True,
+    "depends": [
+        "web",
+        "merp_base",
```
Comment (Tatider) [inline two_factor_otp_auth/__manifest__.py]: Why?

File: two_factor_otp_auth/models/res_users.py (hunk 9 above)
Comment (Tatider) [reply two_factor_otp_auth/models/res_users.py:130]: I'm a little worried that you are using `_`. The same as for translations

File: two_factor_otp_auth/data/ir_actions_server_data.xml [hunk 29]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="1">
```
Comment (Tatider) [inline two_factor_otp_auth/data/ir_actions_server_data.xml]: since version `12.0` `<data>` tag is redundant - you can use `<odoo noupdate="1">` instead

**Review by Tatider — COMMENTED** [review]
Few interesting moments in code

File: merp_base/models/res_config.py (hunk 12 above)
Comment (Werded) [reply merp_base/models/res_config.py]: done

File: two_factor_otp_auth/controllers/main.py (hunk 14 above)
Comment (Werded) [reply two_factor_otp_auth/controllers/main.py]: done

File: two_factor_otp_auth/data/ir_actions_server_data.xml (hunk 15 above)
Comment (Werded) [reply two_factor_otp_auth/data/ir_actions_server_data.xml]: removed groups, be cause in 12.0 Odoo ir.action.server haven't field 'groups_id'

File: two_factor_otp_auth/__manifest__.py (hunk 26 above)
Comment (Werded) [reply two_factor_otp_auth/__manifest__.py]: done

File: two_factor_otp_auth/controllers/main.py (hunk 27 above)
Comment (Werded) [reply two_factor_otp_auth/controllers/main.py]: renamed to '_redirect_to_2fa'.

File: two_factor_otp_auth/__manifest__.py (hunk 28 above)
Comment (Werded) [reply two_factor_otp_auth/__manifest__.py]: removed

File: two_factor_otp_auth/data/ir_actions_server_data.xml (hunk 29 above)
Comment (Werded) [reply two_factor_otp_auth/data/ir_actions_server_data.xml]: removed

File: two_factor_otp_auth/models/res_users.py (hunk 17 above)
Comment (Werded) [reply two_factor_otp_auth/models/res_users.py:150]: done

File: two_factor_otp_auth/models/res_users.py (hunk 18 above)
Comment (Werded) [reply two_factor_otp_auth/models/res_users.py]: done

File: two_factor_otp_auth/models/res_users.py (hunk 19 above)
Comment (Werded) [reply two_factor_otp_auth/models/res_users.py]: added Extra Info

File: two_factor_otp_auth/models/res_users.py (hunk 21 above)
Comment (Werded) [reply two_factor_otp_auth/models/res_users.py]: moved into beginning of file

File: two_factor_otp_auth/models/res_users.py (hunk 22 above)
Comment (Werded) [reply two_factor_otp_auth/models/res_users.py]: done

File: two_factor_otp_auth/security/two_factor_otp_auth.xml (hunk 23 above)
Comment (Werded) [reply two_factor_otp_auth/security/two_factor_otp_auth.xml]: there is no point, removed

File: two_factor_otp_auth/views/res_users_view.xml (hunk 24 above)
Comment (Werded) [reply two_factor_otp_auth/views/res_users_view.xml:13]: done

File: two_factor_otp_auth/views/res_users_view.xml (hunk 25 above)
Comment (Werded) [reply two_factor_otp_auth/views/res_users_view.xml]: done

File: two_factor_otp_auth/models/res_users.py (hunk 20 above)
Comment (Werded) [reply two_factor_otp_auth/models/res_users.py]: I think that method must be 'static' , be cause he need only for checking access rights current  user.

**Pull request merged**